adds storage & query layer to events correlation engine

sbcd90 opened the pull request:
<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
adds hnsw graph based storage & query layer to events correlation engine

### Related Issues
https://github.com/opensearch-project/OpenSearch/issues/6854
<!-- List any other related issues here -->

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [ ] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff
- [X] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

sbcd90 commented:
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/15536/
> * **CommitID:** [156d3c1](https://github.com/opensearch-project/OpenSearch/commit/156d3c16da484d3d4f02b93d279b6221385ceb8d)
>   Please examine the workflow log, locate, and copy-paste the failure(s) below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

build fails for tests
```
Tests with failures:
 - org.opensearch.remotestore.SegmentReplicationRemoteStoreIT.testPressureServiceStats
 - org.opensearch.remotestore.SegmentReplicationRemoteStoreIT.testReplicaHasDiffFilesThanPrimary
```

sbcd90 commented on plugins/events-correlation-engine/src/javaRestTest/java/org/opensearch/plugin/correlation/CorrelationVectorsEngineIT.java:
```diff
@@ -0,0 +1,247 @@
+     * @throws IOException IOException
+     */
+    @SuppressWarnings("unchecked")
+    public void testQuery() throws IOException {
```
added unhappy tests for missing index, corrupt/wrong mapping.

sbcd90 commented on plugins/events-correlation-engine/build.gradle:
```diff
@@ -17,5 +17,7 @@ opensearchplugin {
   classname 'org.opensearch.plugin.correlation.EventsCorrelationPlugin'
 }
 
+forbiddenApis.ignoreFailures = true
```
already fixed it.

andrross commented on plugins/events-correlation-engine/build.gradle:
```diff
@@ -17,5 +17,7 @@ opensearchplugin {
   classname 'org.opensearch.plugin.correlation.EventsCorrelationPlugin'
 }
 
+// forbiddenApis.ignoreFailures = true
```
Please remove

sbcd90 replied:
removed it.

andrross reviewed (commented):
Lot of test failures here. Can you rebase the latest from main?

andrross commented on plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/core/index/query/CorrelationQueryBuilder.java:
```diff
@@ -0,0 +1,334 @@
+            this.vector = sin.readFloatArray();
+            this.k = sin.readInt();
+            this.filter = sin.readOptionalNamedWriteable(QueryBuilder.class);
+        } catch (IOException ex) {
```
Why catch this as opposed to letting it propagate?

sbcd90 replied:
fixed it.

sbcd90 commented:
hi @andrross , rebased the pr. all the tests pass now.

getsaurabh02 commented on plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/core/index/codec/correlation950/CorrelationCodec.java:
```diff
@@ -0,0 +1,46 @@
+ * @opensearch.internal
+ */
+public class CorrelationCodec extends FilterCodec {
+    private static final CorrelationCodecVersion VERSION = CorrelationCodecVersion.V_9_5_0;
```
Should we extract out the codec version as part of the config or to a single owner, as its being used across enums and CorrelationCodec definitions.

getsaurabh02 commented on plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/core/index/mapper/VectorFieldMapper.java:
```diff
@@ -0,0 +1,399 @@
+        List<Float> vector = new ArrayList<>();
+        XContentParser.Token token = context.parser().currentToken();
+        float value;
+        if (token == XContentParser.Token.START_ARRAY) {
```
Good to have tests for field mapper logic of value extraction.

getsaurabh02 reviewed (approved):
Thanks @sbcd90 LGTM. Added few comments which seems non-blocking and can be addressed as follow-up.

andrross commented on plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/core/index/codec/BasePerFieldCorrelationVectorsFormat.java:
```diff
@@ -0,0 +1,102 @@
+import java.util.function.Supplier;
+
+/**
+ * Class to define the hyper-parameters m and ef_construction for insert and store of correlation vectors into HNSW graphs based lucene index.
```
Please add `@opensearch.internal` to all new classes not intended for extension, or `@opensearch.api`/`@opensearch.experimental` if it is intended for extension by other plugins.

andrross reviewed (commented):
Mostly looks good to me.

> 56.53% of diff hit (target 70.63%)

The code coverage target is a bit arbitrary but can we get better coverage here from the unit tests?

andrross commented:
@sbcd90 Just a friendly ping on [my comments from last week](https://github.com/opensearch-project/OpenSearch/pull/7350#pullrequestreview-1444772398).

andrross commented:
@sbcd90 Can you link [the meta issue](#6854) in the description as well and be sure to keep that issue up to date with your progress?

sbcd90 commented:
hi @andrross , thanks a lot for the reminders. will look into your comments.

sbcd90 commented:
hi @andrross , improved the code coverage for patch to `67.66%`. will create a separate issue to improve the `code coverage`.
addressed all review comments as well.

andrross commented on plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/core/index/VectorField.java:
```diff
@@ -0,0 +1,38 @@
+        try {
+            final byte[] floatToByte = CorrelationVectorSerializer.floatToByteArray(value);
+            this.setBytesValue(floatToByte);
+        } catch (Exception ex) {
```
`CorrelationVectorSerializer.floatToByteArray` already catches IOExceptions and wraps them as an OpenSearchException. This class then catches the OpenSearchException and wraps it in a RuntimeException. Why?

I recommend that a wrap the IOException in a single place, _and_ that you do not bring in OpenSearchException and/or ExceptionHelpers and instead throw a base Java exception (maybe just RuntimeException) unless you have a good reason to do something else.

sbcd90 replied:
throwing `java.lang.RuntimeException` only now.

andrross replied:
Don't catch `Exception` here. The risk is that if this code is refactored such that `floatToByteArray` where to throw another checked exception, this would swallow it which may or may not be appropriate. Namely, if a blocking call were added that threw an `InterruptedException` then this would code would swallow the interrupt without resetting the interrupt status. Just catch `IOException` here since that is the only checked exception being thrown.

andrross commented on plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/core/index/serializer/CorrelationVectorSerializer.java:
```diff
@@ -0,0 +1,52 @@
+ *
+ * @opensearch.internal
+ */
+public class CorrelationVectorSerializer {
```
Do you need a separate class to hold these functions? Why can't `floatToByteArray` be a function inside VectorField since that appears to be the only place it is used?

`byteToFloatArray` seems to only be used in tests, so can it be moved to a test utility?

sbcd90 replied:
moved `floatToByteArray` to `VectorField` class & moved `byteToFloatArray` to tests.

sbcd90 commented on plugins/events-correlation-engine/build.gradle:
```diff
@@ -19,3 +19,16 @@ opensearchplugin {
+    classDirectories.setFrom(files(classDirectories.files.collect {
+      fileTree(dir: it, exclude: [
+        "org/opensearch/plugin/correlation/EventsCorrelationPlugin*",
+        "org/opensearch/plugin/correlation/core/index/mapper/VectorFieldMapper*"
```
removed it.

sbcd90 commented:
hi @andrross , thanks a lot for reviewing the pr again. addressed all the comments. please have a look.

andrross commented on plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/core/index/CorrelationParamsContext.java:
```diff
@@ -0,0 +1,166 @@
+        }
+        if (contextMap.containsKey(PARAMETERS)) {
+            Object value = contextMap.get(PARAMETERS);
+            if (value == null) {
+                parameters = null;
+            }
```
Why do this? The next line asserts `value` is a `Map` or else throws. Am I missing something here?

sbcd90 replied:
removed the line.

andrross commented on plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/core/index/CorrelationParamsContext.java:
```diff
@@ -0,0 +1,166 @@
+        return builder;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        CorrelationParamsContext that = (CorrelationParamsContext) o;
+        return similarityFunction == that.similarityFunction && parameters.equals(that.parameters);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(similarityFunction, parameters);
+    }
```
Do you need equals and hashcode? The `parameters` map has values of type `Object` so you have no way of knowing whether those types will use identity based equals and hash code, which can lead to seemingly identical instances not being equal. I would recommend to either remove these and use identity-based equality for this class, or else constrain the value type of `parameters` to a type that implements proper equals/hash code.

sbcd90 replied:
removed `equals` & `hashCode` method.

sbcd90 commented:
hi @andrross , fixed it. https://github.com/opensearch-project/OpenSearch/pull/7350#discussion_r1235489557
Addressed all the comments. please have a look.
